### ISSUE TITLE
[PATCH] WGI: Fix HSTRING memory leak.

### ISSUE THREAD

**Issue by SDLBugzilla** [libsdl-org/SDL#3902]
# This bug report was migrated from our old Bugzilla tracker.

These attachments are available in the static archive:

* [Release the HSTRING to fix memleak. (sdl2_wgi_hstring_leak.patch, text/plain, 2020-12-01 23:49:42 +0000, 2514 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=4551)

**Reported in version:** 2.0.13
**Reported for operating system, platform:** Windows 10, All

# Comments on the original bug report:

On 2020-12-01 23:49:42 +0000, Joel Linn wrote:

> Created attachment 4551
> Release the HSTRING to fix memleak.
> 
> I might be wrong about this one since the documentation on the WinRT C interface is pretty much non-existent.
> 
> However I am pretty certain the property-getter transfers ownership of the HSTRING to the caller or rather increments the reference count. Anyways, we shall release the HSTRING.

On 2020-12-09 14:28:36 +0000, Sam Lantinga wrote:

> I'm not sure about this one, so I'm going to wait to include it until after 2.0.14 release.

On 2020-12-09 23:31:38 +0000, Joel Linn wrote:

> TLDR; https://godbolt.org/z/43fd8G
> 
> Let's deduce this from C++ reference code:
> 
> https://docs.microsoft.com/en-us/cpp/cppcx/wrl/how-to-activate-and-use-a-windows-runtime-component-using-wrl?view=msvc-160
> At the bottom of the page there is this snippet:
> ```
> int wmain()
> {
>     /* ... more code ... */
> 
>     // Get the domain part of the URI.
>     HString domainName;
>     hr = uri->get_Domain(domainName.GetAddressOf());
>     if (FAILED(hr))
>     {
>         return PrintError(__LINE__, hr);
>     }
> 
>     // Print the domain name and return.
>     wprintf_s(L"Domain name: %s\n", domainName.GetRawBuffer(nullptr));
> 
>     // All smart pointers and RAII objects go out of scope here.
> }
> ```
> 
> `HString` is defined in `corewrappers.h` and the call chain for the destructor is:
> `~HString() -> Release() -> ::WindowsDeleteString()`
> 
> QED

On 2020-12-10 04:29:24 +0000, Sam Lantinga wrote:

> Sold!
> https://hg.libsdl.org/SDL/rev/e4681c33144e
> 
> Thanks for researching this. :)
